File: APIApplication/src/main/java/org/apiapplication/services/interfaces/GroupService.java
package org.apiapplication.services.interfaces;

import org.apiapplication.dto.common.IdDto;
import org.apiapplication.dto.group.*;

import java.util.List;

public interface GroupService {
    List<GroupDto> get();

    GroupDto getById(int id);

    IdDto add(AddGroupDto addGroupDto);

    void update(int groupId, UpdateGroupDto updateGroupDto);

    void delete(int groupId);

    void addStudents(int groupId, SetStudentsDto setStudentsDto);

    void removeStudents(int groupId, SetStudentsDto setStudentsDto);

    void addSubjects(int groupId, SetSubjectsDto setSubjectsDto);

    void removeSubjects(int groupId, SetSubjectsDto setSubjectsDto);
}

File: FrontendApplication/frontend-project/src/app/shared/services/group.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { baseUrl } from "../constants/url.constant";
import { IdDto } from "../models/common.model";
import { GroupDto, AddGroupDto, UpdateGroupDto } from "../models/group.model";

@Injectable({
    providedIn: 'root'
})
export class GroupService {
    private url: string = baseUrl + "/groups";

    constructor(private readonly httpClient: HttpClient) {
    }

    getAll(): Observable<GroupDto[]> {
        return this.httpClient.get<GroupDto[]>(`${this.url}`);
    }

    getById(id: number): Observable<GroupDto> {
        return this.httpClient.get<GroupDto>(`${this.url}/${id}`);
    }

    add(dto: AddGroupDto): Observable<IdDto> {
        return this.httpClient.post<IdDto>(`${this.url}`, dto);
    }

    update(id: number, dto: UpdateGroupDto): Observable<any> {
        return this.httpClient.put<any>(`${this.url}/${id}`, dto);
    }

    setStudents(id: number, userIds: number[]): Observable<any> {
        return this.httpClient.put<any>(`${this.url}/${id}/setStudents`, userIds);
    }

    setSubjects(id: number, subjectIds: number[]): Observable<any> {
        return this.httpClient.put<any>(`${this.url}/${id}/setSubjects`, subjectIds);
    }

    delete(id: number): Observable<any> {
        const queryParams = new HttpParams();
        queryParams.set('id', id);

        return this.httpClient.delete<any>(`${this.url}/${id}`, { params: queryParams });
    }
}